### PR TITLE
Move frobenius_inner_product to qml.math

### PR DESCRIPTION
**Pull request by johannesjmeyer**
**Context:**
In the QML demo (https://github.com/PennyLaneAI/qml/pull/266), it turned out that the function `frobenius_inner_product` was not differentiable because regular numpy was imported in `qml.utils`.

**Description of the Change:**
`frobenius_inner_product` was moved to `qml.math`. To this end, new files were introduced.

**Benefits:**
Differentiability.

**Possible Drawbacks:**
Slight decrease in performance.

### CODE REVIEW

File: pennylane/math/classical.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    Returns:
+        float: Frobenius inner product of A and B
+
+    **Example**
+
+    >>> A = np.random.random((3,3))
+    >>> B = np.random.random((3,3))
+    >>> qml.math.frobenius_inner_product(A, B)
+    3.091948202943376
+    """
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/math/classical.py:43]: Tangential thought: we should provide a wrapper function to automate this:

```python
def something_descriptive(values):
    return np.coerce(values, like=_multi_dispatch(values))
```

[the remark this answers is not in the document]
Comment (johannesjmeyer) [reply pennylane/math/classical.py:43]: Cool, had no clue how that works! But that is very nice functionality :)

File: tests/math/test_basic_math.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+                np.array([[0.7, -7.3], [-1.0, -2.9]]),
+                False,
+                -21.75,
+            ),
+            (np.eye(2), np.eye(2), True, 1.0),
+            (
+                np.array([[1.0, 2.3], [-1.3, 2.4]]),
+                np.array([[0.7, -7.3], [-1.0, -2.9]]),
+                True,
+                -0.7381450594,
```
[the remark this answers is not in the document]
Comment (johannesjmeyer) [reply tests/math/test_basic_math.py]: I think it would be great if you added a way to the library to automate this, that can't be that hard, right? I.e. run the same test case with different variable types depending on which libraries are available.

File: tests/math/test_basic_math.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+            ),
+        ],
+    )
+    def test_frobenius_inner_product(self, A, B, normalize, expected):
+        assert expected == pytest.approx(fn.frobenius_inner_product(A, B, normalize=normalize))
```
[the remark this answers is not in the document]
Comment (johannesjmeyer) [reply tests/math/test_basic_math.py:109]: Actually I have no clue how to write the gradient test, can you help me out? I tried something like
```python
def test_frobenius_inner_product_gradient(self):
        A = np.array([[1.0, 2.3], [-1.3, 2.4]])
        B = torch.autograd.Variable(torch.randn(2, 2).type(tf.complex128), requires_grad=True)
        dB = np.ones((2, 2), dtype=float)
        g = qml.grad(fn.frobenius_inner_product(A, B))

        assert g(B) == 2 * fn.frobenius_inner_product(A, dB)
```
which didn't really work...

[the remark this answers is not in the document]
Comment (josh146) [reply tests/math/test_basic_math.py:109]: @johannesjmeyer it looks like you are using `qml.grad` with pytorch?

`qml.grad` only wraps `autograd.grad`, so you’ll need to use the torch equivalent method of doing the gradient:

```python
result = fn.frobenius_inner_product(A, B)
result.backward()
grad = B.grad
```

[the remark this answers is not in the document]
Comment (johannesjmeyer) [reply tests/math/test_basic_math.py:111]: Thanks!

**Review by josh146 — COMMENTED** [review]
Hey @johannesjmeyer, just double checking the status of this PR 🙂 

Looks like it is in good shape, it is just missing tests for TF, autograd, and JAX gradients?

**Comment by johannesjmeyer** [comment]
I already added the gradient tests and re-requested a review before, can you have a look if the tests are alright? I'm on vacation until August 2nd.

**Review by josh146 — APPROVED** [review]
Apologies for the delay @johannesjmeyer; looks good from my end 💯 

In my previous comment I think I was referring to the fact that there is a Torch gradient test, but no gradient test for JAX/TF/Autograd. However, from looking at the code again, I am satisfied that the Torch test is sufficient (the amount of processing is minimal!)

**Comment by johannesjmeyer** [comment]
Cool, then you can merge it in :) I don't have the sufficient rights anymore.

**Comment by co9olguy** [comment]
@josh146 worth a changelog update? (unless I missed it)

**Comment by josh146** [comment]
@co9olguy good catch. It was decided to simply amend the Kernel module changelog entry in an earlier review, but this will have to be updated now that it is crossing a release.

**Comment by johannesjmeyer** [comment]
Should I take care of that?

**Comment by josh146** [comment]
> Should I take care of that?

All good @johannesjmeyer, I can take this one! My fault for the PR delay 🙂
